src/mapgen/mapgen.cpp

```diff
@@ -980,39 +981,65 @@ void MapgenBasic::generateDungeons(s16 max_stone_y)
 ////
 
 GenerateNotifier::GenerateNotifier(u32 notify_on,
-	const std::set<u32> *notify_on_deco_ids)
+	const std::set<u32> *notify_on_deco_ids,
+	const std::set<std::string> *notify_on_custom)
 {
 	m_notify_on = notify_on;
 	m_notify_on_deco_ids = notify_on_deco_ids;
+	m_notify_on_custom = notify_on_custom;
 }
 
 
-bool GenerateNotifier::addEvent(GenNotifyType type, v3s16 pos, u32 id)
+bool GenerateNotifier::addEvent(GenNotifyType type, v3s16 pos)
 {
-	if (!(m_notify_on & (1 << type)))
+	assert(type != GENNOTIFY_DECORATION && type != GENNOTIFY_CUSTOM);
+	if (!notifyOn(type))
 		return false;
 
-	if (type == GENNOTIFY_DECORATION &&
-		m_notify_on_deco_ids->find(id) == m_notify_on_deco_ids->cend())
+	GenNotifyEvent gne;
+	gne.type = type;
+	gne.pos  = pos;
+	m_notify_events.emplace_back(std::move(gne));
+	return true;
+}
+
+
+bool GenerateNotifier::addDecorationEvent(v3s16 pos, u32 id)
+{
+	if (!notifyOn(GENNOTIFY_DECORATION))
+		return false;
+	assert(m_notify_on_deco_ids);
+	if (m_notify_on_deco_ids->find(id) == m_notify_on_deco_ids->cend())
 		return false;
 
 	GenNotifyEvent gne;
-	gne.type = type;
+	gne.type = GENNOTIFY_DECORATION;
 	gne.pos  = pos;
 	gne.id   = id;
-	m_notify_events.push_back(gne);
+	m_notify_events.emplace_back(std::move(gne));
+	return true;
+}
+
+
+bool GenerateNotifier::setCustom(const std::string &key, const std::string &value)
+{
+	if (!notifyOn(GENNOTIFY_CUSTOM))
+		return false;
+	assert(m_notify_on_custom);
+	if (m_notify_on_custom->count(key) == 0)
```

Comment:
This function should allow overwriting just like all other setter functions that I can think of. Someone might want to process and replace the data from another mapgen env (daisy-chain?).

Please correct me in case I misunderstood the concept of gen notify data.

Reply:
Yeah I don't know why I coded it like this.